### PR TITLE
[DPE-3374] Add authorization servlet filter logic

### PR DESCRIPTION
**Pull request by welpaolo**
This PR introduce the authorization servlet filter:
- Servlet filter activated when oathkeeper is related 
- New configuration option
- New integration tests

This PR superseded #32

### CODE REVIEW

**Review by theoctober19th — APPROVED** [review]
LGTM

**Review by deusebio — CHANGES_REQUESTED** [review]
The code looks fine for the most part!

there is something fuzzy in the implementation of

`def _on_auth_proxy_changed(self, _: AuthProxyRelationRemovedEvent)`

we need to double check and get consolidated. Please check my comments. Once that is addressed, I believe we can go ahead and merge the PR

I also have a suggestion related to a point of @theoctober19th as well, but this is non-blocking and up to you.

File: src/events/ingress.py [hunk 1]
```diff
@@ -72,10 +80,29 @@ def _on_ingress_revoked(self, _: IngressPerAppRevokedEvent):
...
+            self.context.s3, self.context.ingress, None, self.context.authorized_users
+        )
 
         self.charm.unit.status = self.get_app_status(self.context.s3, self.context.ingress, None)
         if self.charm.unit.is_leader():
             self.charm.app.status = self.get_app_status(
                 self.context.s3, self.context.ingress, None
             )
+
+    @compute_status
```
[the remark this answers is not in the document]
Comment (welpaolo) [reply src/events/ingress.py:93]: Sorry, my bad I was playing around with the decorator to test and I left both the decorator and the update logic there... I am leaving the decorator and remove the update logic.

File: src/events/ingress.py [hunk 2]
```diff
@@ -72,10 +80,29 @@ def _on_ingress_revoked(self, _: IngressPerAppRevokedEvent):
...
+    def _on_auth_proxy_changed(self, _: AuthProxyRelationRemovedEvent):
```
[the remark this answers is not in the document]
Comment (welpaolo) [reply src/events/ingress.py:94]: Type of the event updated

File: src/events/ingress.py [hunk 3]
```diff
@@ -72,10 +80,29 @@ def _on_ingress_revoked(self, _: IngressPerAppRevokedEvent):
...
+        """Handle the removal of the AuthProxy."""
+        self.logger.info("AuthProxy configuration gone")
+        self.history_server.update(
+            self.context.s3,
+            self.context.ingress,
+            self.context.auth_proxy_config,
```
[the remark this answers is not in the document]
Comment (welpaolo) [reply src/events/ingress.py:100]: This block of code will handle the change of configuration

File: src/events/ingress.py [hunk 4]
```diff
@@ -72,10 +79,24 @@ def _on_ingress_revoked(self, _: IngressPerAppRevokedEvent):
     def _on_auth_proxy_removed(self, _: AuthProxyRelationRemovedEvent):
         """Handle the removal of the AuthProxy."""
         self.logger.info("AuthProxy configuration gone")
-        self.history_server.update(self.context.s3, self.context.ingress)
+        self.history_server.update(
+            self.context.s3, self.context.ingress, self.context.authorized_users
```
Comment (deusebio) [inline src/events/ingress.py:83]:
```suggestion
            self.context.s3, self.context.ingress, None
```

shouldn't this be none?

**Review by deusebio — APPROVED** [review]

**Comment by deusebio** [comment]
Looks great!

**Pull request merged**